docs/VARIABLE_ROLES.md

```diff
@@ -0,0 +1,112 @@
+# Variable Roles in the Basic Model Interface (BMI)
+
+Here, we use the term *variable* very broadly (as a “catch all”) for any type of variable that appears in a computer model.
+The purpose of this document is to define a set of distinct and non-overlapping *roles* that a variable can play in a model.
+Introducing variable roles allow support for tasks such as serialization and calibration that require working with some subset of all model variables.
+Note that a variable’s role may sometimes change based on a setting or option in the model’s configuration file.
+For example, some variables may only be computed and made available as outputs when a certain option is activated.
+
+0. **all**:   This special role name is used to refer to the set of all model variables, except for purely local variables that are only initialized and used within a function (like a loop counter) but are neither an argument nor return value of that function.
+This is the set of variables that must be serialized if we wish to save the complete *state* of the computer model.
+(Here, we are thinking of *state* in the computer science sense, as “remembered information” or a “stateful system”.)

```

Comment:
Purely a wording suggestion, take it or leave it:

(Here, we are referring to *state* in ...

OR

(Here, we are using *state* in ...

Reply:
i'll take it. thx